enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,971 @@
+Most applications will set something like this for static assets:
+```sh
+cache-control: max-age=7200, must-revalidate, public
+```
+But the OpenShift router (HAProxy) will add cache-control: private which results in both headers in the response:
+```sh
+cache-control: private
+cache-control: max-age=7200, must-revalidate, public
+```
+This leads to issues with caching: CDNs like Cloudflare will not cache your files.
+
+- Can we please make sure to only add the `cache-control: private` header if there is no `cache-control: public` header already present in the response?
+  Answer: A common technique is to put static assets on a different virtual host. In OpenShift terms, that means using a distinct route for static assets. This route can
+  be annotated with `haproxy.router.openshift.io/disable_cookies=true` to prevent setting the cookie and thus prevent setting cache-control.
+
+- Would we be able to [Replace a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-a-header-by-using-a-regular-expression) or [Replace part of a header by using a regular expression](https://www.haproxy.com/documentation/hapee/latest/traffic-routing/rewrites/rewrite-responses/#replace-part-of-a-header-by-using-a-regular-expression)?
+  Answer: No, this is out of scope for this enhancement. Action `Set` will perform `set-header` in the haproxy.config for `http-response` and `http-request` which will either set the value if the value is not present or replace the existing value.
+
+- Should this enhancement allow setting dynamic header values?
+  Answer: Yes, you would be able to set a minimal set of dynamic header values. We had three options for this EP.
+    1. Allow dynamic header values with all sample fetchers and converters.
+    2. Allow a minimal set of dynamic values that we can confidently validate, focusing on the things that people have asked for on the RFE or [EP](https://github.com/openshift/enhancements/pull/1296#issuecomment-1396103783).
+    3. Prohibit dynamic values entirely.
+    However, as few customers asked for it, we went ahead with option number 2.
+    For setting dynamic header values for HTTP responses, allowed sample fetchers are `res.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
+    For setting dynamic header values for HTTP requests, allowed sample fetchers are `req.hdr` and `ssl_c_der`. Allowed converters are `lower` and `base64`.
+    Example: One might be able to set value of `x-forwarded-client-cert` to `%{+Q}[ssl_c_der,base64]`.
+
+- Certain headers like HSTS are not allowed to be set using this API as there is already a route annotation available for this.
```

Comment:
Again, HSTS is not just a route annotation.  It is a setting that an admin can add to a domain, via `spec.requiredHSTSPolicies`, as explained in https://docs.openshift.com/container-platform/4.13/networking/routes/route-configuration.html#nw-enforcing-hsts-per-domain_route-configuration.